R/germline.R

```diff
@@ -158,10 +173,95 @@ merge_reference_sequences <- function(data, chain_letter, species, sample_name)
   if (nrow(data) == 0) {
     stop(
       "After merging with reference, the data ",
-      optional_from_sample(sample_name),
+      optional_sample("from sample ", sample_name, " "),
       "is empty.\n",
-      "There were no valid alleles, or the data was initially empty."
+      "There were no valid alleles in the data!"
+    )
+  }
+  return(data)
+}
+
+validate_genes_edges <- function(data, sample_name) {
+  if (nrow(data) == 0) {
+    stop(
+      "Sample ",
+      optional_sample("", sample_name, " "),
+      "dataframe is empty!"
+    )
+  }
+  for (column in c("V.end", "J.start")) {
+    if (!(column %in% colnames(data))) {
+      stop(
+        "Missing mandatory ",
+        column,
+        " column",
+        optional_sample(" in sample ", sample_name, ""),
+        "!"
+      )
+    }
+    if (all(is.na(data[, column]))) {
+      stop(
+        "No data in mandatory ",
+        column,
+        " column",
+        optional_sample(" in sample ", sample_name, ""),
+        "!"
+      )
+    }
+  }
+  old_length <- nrow(data)
+  data %<>% drop_na(V.end, J.start)
+  dropped_num <- old_length - nrow(data)
+  if (dropped_num > 0) {
+    warning(
+      dropped_num,
+      " rows from ",
+      old_length,
+      optional_sample(" in sample ", sample_name, ""),
+      " were dropped because of missing values in mandatory columns V.end and J.start!"
     )
   }
+  if (nrow(data) == 0) {
+    stop(
+      "Sample ",
+      optional_sample("", sample_name, " "),
+      "dataframe is empty after dropping missing values!"
+    )
+  }
+  return(data)
+}
+
+validate_chains_length <- function(data, sample_name, min_good_length = 5) {
```

Comment:
Maybe we can find a better name for `min_good_length` parameter? And it is worth adding a comment to describe it